### DISCUSSION TITLE
Docker very high memory utilization

### DISCUSSION THREAD

**Discussion by Kaesewuerfel** [pterodactyl/panel#2981] [General Community]
Hello,

i use pterodactyl and I host 28 minecraft servers (1 GB Ram) on a 60 GB Ram System and the docker.sock is using 13 GB of my Ram.
It's pretty much. How can I reduce this?

**Comment by parkervcp** [comment]
So each server is using around .75 gb of ram? I think you are just seeing the java processes memory showing under docker.

**Reply by Kaesewuerfel** [reply]
Yes, java and docker is using ram.
![image](https://user-images.githubusercontent.com/74424677/103785941-15447580-503c-11eb-86f9-c0173f6a0507.png)

**Reply by matthewpi** [reply]
Can you enable Tree (`F5`) and send a screenshot?

**Reply by schrej** [reply]
As far as I know, the processes within docker containers are not children of dockerd. Therefore their memory is not counted towards dockerd. So this could either be caused by docker itself, or by us inadequately using it.
Could you try to restart wings and see if that reduces memory usage?

**Reply by LordofAgents** [reply]
First, i am a colleague of @Kaesewuerfel , so we are talking about the same server and system.

> As far as I know, the processes within docker containers are not children of dockerd. Therefore their memory is not counted towards dockerd. So this could either be caused by docker itself, or by us inadequately using it.
> Could you try to restart wings and see if that reduces memory usage?

Did it, nothing changed. Exactly same usage of memory. So this may be caused of docker itself. But the memory usage of wings reduced, but wings doesnt need much.

> Can you enable Tree (`F5`) and send a screenshot?

![1](https://user-images.githubusercontent.com/52619285/103796349-6017ba80-5047-11eb-8323-e1e9dff21973.png)
Then many docker-proxy-processes and then:
![Screenshot_20210106_174248](https://user-images.githubusercontent.com/52619285/103796399-6c9c1300-5047-11eb-876b-9c6397cd89a2.png)

Do you need more details?

**Reply by schrej** [reply]
All right, looks like there are tons of docker-proxy instances running which consume tons of memory and are likely the cause of this.

According to [this article](https://windsock.io/the-docker-proxy/) (admittedly quite old, but I think still true), the proxy is only really used when nothing better like iptables is available. Do you by chance have ip_forwarding disabled on your system?
In general I don't think this is caused by Pterodactyl, but by an incorrect docker setup.

Another interesting article: https://www.ipspace.net/kb/DockerSvc/40-userland-proxy.html

**Reply by LordofAgents** [reply]
Okay.... First thank you for the articles, this was very helpful. We have three nodes, all are equal. 
There was ufw installed on all nodes (i know, beginner mistake and program), uninstalling and reboot solved the problem for 2/3 nodes. I'm investigating, why this doesn't help on the last node.

**Reply by Kaesewuerfel** [reply]
now it's again on 2 nodes :/

**Reply by LordofAgents** [reply]
We monitor the nodes, but I think we solved it!
Added --userland-proxy:false and --iptables:true to the Options for Docker. ufw is uninstalled. We have no docker-proxy Processes anymore. Works great for now, we're waiting until we got more load.

Thanks for your help!
